### PR TITLE
Ignore unknown root elements starting with `x-`

### LINKED ISSUE #1246: Support YAML syntax features: Anchors, Aliases and Extensions
### Checklist for this issue

- [x] 🚨review the [guidelines for contributing](../blob/master/docs/CONTRIBUTING.md) to this repository.

### Feature Request

Would be nice to have native support for situations like this

```yaml
x-ec2_anchor: &ec2_anchor
  description: "ec2-anchor"
  labelString: "ec2-anchor"
  type: "ec2-anchor"
  ami: "ec2-anchor"
  amiType:
    unixData:
      sshPort: "22"
  associatePublicIp: false
  connectBySSHProcess: false
  connectionStrategy: "PRIVATE_IP"
  deleteRootOnTermination: true
  ebsOptimized: false
  idleTerminationMinutes: "2"
  initScript: ""
  maxTotalUses: 1
  minimumNumberOfInstances: 0
  instanceCapStr: 50
  iamInstanceProfile: ""
  jvmopts: ""
  mode: "NORMAL"
  monitoring: false
  numExecutors: 1
  remoteAdmin: "ubuntu"
  remoteFS: "/home/ubuntu"
  spotConfig:
    useBidPrice: false
    fallbackToOndemand: true
    spotBlockReservationDurationStr: 1
  stopOnTerminate: false
  securityGroups: "default"
  subnetId: "subnet-XXXXXX"
  t2Unlimited: false
  useDedicatedTenancy: false
  useEphemeralDevices: false
  userData: ""
  zone: ""
  tags:
    - name: "label"
      value: "jenkins-slave"
    - name: "environment"
      value: "development"
    - name: "cost_center"
      value: "devinfra"

jenkins:
  clouds:
    - amazonEC2:
        cloudName: "us-east-common"
        useInstanceProfileForCredentials: true
        instanceCapStr: 200
        noDelayProvisioning: true
        privateKey: ${EC2_PRIVATE_KEY}
        region: "us-east-1"
        templates:
          - <<: *ec2_anchor
            ami: "ami-XXXXXX"
            type: "T3Small"
            description: "T3Small:xenial"
            labelString: "ubuntu:xenial aws:small"
          - <<: *ec2_anchor
            ami: "ami-XXXXXX"
            type: "T3Large"
            description: "T3Large:xenial"
            labelString: "ubuntu:xenial aws:large"
          - <<: *ec2_anchor
            ami: "ami-YYYYYY"
            type: "T3Small"
            description: "T3Small:bionic"
            labelString: "ubuntu:bionic aws:small"
          - <<: *ec2_anchor
            ami: "ami-YYYYYY"
            type: "T3Large"
            description: "T3Large:bionic"
            labelString: "ubuntu:bionic aws:large"
```

Currently, this config causes such exception:

<details>

<summary>Stack Trace (clickable)</summary>

```
io.jenkins.plugins.casc.ConfiguratorException: No configurator for root element <x-ec2_anchor>
	at io.jenkins.plugins.casc.ConfigurationAsCode.invokeWith(ConfigurationAsCode.java:710)
	at io.jenkins.plugins.casc.ConfigurationAsCode.checkWith(ConfigurationAsCode.java:746)
	at io.jenkins.plugins.casc.ConfigurationAsCode.configureWith(ConfigurationAsCode.java:731)
	at io.jenkins.plugins.casc.ConfigurationAsCode.configureWith(ConfigurationAsCode.java:611)
	at io.jenkins.plugins.casc.ConfigurationAsCode.configure(ConfigurationAsCode.java:292)
	at io.jenkins.plugins.casc.ConfigurationAsCode.doReload(ConfigurationAsCode.java:156)
	at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627)
	at org.kohsuke.stapler.Function$MethodFunction.invoke(Function.java:396)
	at org.kohsuke.stapler.Function$InstanceFunction.invoke(Function.java:408)
	at org.kohsuke.stapler.interceptor.RequirePOST$Processor.invoke(RequirePOST.java:77)
	at org.kohsuke.stapler.PreInvokeInterceptedFunction.invoke(PreInvokeInterceptedFunction.java:26)
	at org.kohsuke.stapler.Function.bindAndInvoke(Function.java:212)
	at org.kohsuke.stapler.Function.bindAndInvokeAndServeResponse(Function.java:145)
	at org.kohsuke.stapler.MetaClass$11.doDispatch(MetaClass.java:535)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:747)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:878)
	at org.kohsuke.stapler.MetaClass$9.dispatch(MetaClass.java:456)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:747)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:878)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:676)
	at org.kohsuke.stapler.Stapler.service(Stapler.java:238)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:873)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1623)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:154)
	at com.cloudbees.jenkins.support.slowrequest.SlowRequestFilter.doFilter(SlowRequestFilter.java:37)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at jenkins.metrics.impl.MetricsFilter.doFilter(MetricsFilter.java:125)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at jenkins.telemetry.impl.UserLanguages$AcceptLanguageFilter.doFilter(UserLanguages.java:128)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at hudson.util.PluginServletFilter.doFilter(PluginServletFilter.java:157)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
	at hudson.security.csrf.CrumbFilter.doFilter(CrumbFilter.java:64)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:84)
	at hudson.security.UnwrapSecurityExceptionFilter.doFilter(UnwrapSecurityExceptionFilter.java:51)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:117)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.providers.anonymous.AnonymousProcessingFilter.doFilter(AnonymousProcessingFilter.java:125)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.rememberme.RememberMeProcessingFilter.doFilter(RememberMeProcessingFilter.java:142)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.AbstractProcessingFilter.doFilter(AbstractProcessingFilter.java:271)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.BasicHeaderProcessor.doFilter(BasicHeaderProcessor.java:93)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.context.HttpSessionContextIntegrationFilter.doFilter(HttpSessionContextIntegrationFilter.java:249)
	at hudson.security.HttpSessionContextIntegrationFilter2.doFilter(HttpSessionContextIntegrationFilter2.java:67)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at hudson.security.ChainedServletFilter.doFilter(ChainedServletFilter.java:90)
	at hudson.security.HudsonFilter.doFilter(HudsonFilter.java:171)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
	at org.kohsuke.stapler.compression.CompressionFilter.doFilter(CompressionFilter.java:49)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
	at hudson.util.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:82)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
	at org.kohsuke.stapler.DiagnosticThreadNameFilter.doFilter(DiagnosticThreadNameFilter.java:30)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:540)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:146)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:524)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:257)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1700)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1345)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:480)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1667)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1247)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:505)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:370)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:267)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:698)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:804)
	at java.lang.Thread.run(Thread.java:748)
```

</details>

### Relevant links:
- [jenkinsci/configuration-as-code-plugin/demos/ec2](https://github.com/jenkinsci/configuration-as-code-plugin/tree/master/demos/ec2)
- [medium/anchors-aliases-extensions](https://medium.com/@kinghuang/docker-compose-anchors-aliases-extensions-a1e4105d70bd) (docker-compose)

### PR DESCRIPTION
**Pull request by jetersen**
This should help #1246 to use anchors on root level.

Added tests for unknown root elements and added an alias test for EC2 cloud to demonstrate the potential.

cc @v1v @timja @almiroshnich 

fixes #1246 

### Your checklist for this pull request

🚨 Please review the [guidelines for contributing](../blob/master/docs/CONTRIBUTING.md) to this repository.

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side) and not your master branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or in [Jenkins JIRA](https://issues.jenkins-ci.org)
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Did you provide a test-case? That demonstrates feature works or fixes the issue.

### CODE REVIEW

File: plugin/src/main/java/io/jenkins/plugins/casc/ConfigurationAsCode.java [hunk 1]
```diff
@@ -712,11 +712,24 @@ private static void invokeWith(Mapping entries, ConfiguratorOperation function)
...
+                }
+            });
+
+            if (!unknownKeys.isEmpty()) {
+                throw new ConfiguratorException(format("No configurator for the following root elements %s", String.join(", ", unknownKeys)));
+            }
         }
     }
 
+    static boolean ignoreAliasEntry(String key) {
```
Comment (jetersen) [inline plugin/src/main/java/io/jenkins/plugins/casc/ConfigurationAsCode.java:729]: Naming is hard 😓 But at least I wanted to give the condition a name.

File: integrations/pom.xml [hunk 2]
```diff
@@ -99,7 +99,7 @@
     <dependency>
       <groupId>org.jenkins-ci.plugins</groupId>
       <artifactId>config-file-provider</artifactId>
-      <version>3.4.1</version>
+      <version>3.6.3</version>
```
Comment (jetersen) [inline integrations/pom.xml:102]: Could be done in a separate PR 😅

File: plugin/src/main/java/io/jenkins/plugins/casc/ConfigurationAsCode.java (hunk 1 above)
Comment (timja) [reply plugin/src/main/java/io/jenkins/plugins/casc/ConfigurationAsCode.java:729]: This condition seems the wrong way round to me from the name, isNotAliasEntry?

**Comment by jetersen** [comment]
@almiroshnich any suggestion on how to document this feature?

**Comment by almiroshnich** [comment]
@jetersen huge kudos to you for doing this 🚀 

As for the documentation subject, I don't feel myself competent enough here to make suggestions, but if you'd ask me directly - I'd put a note about such functionality under [jenkinsci/configuration-as-code-plugin//README.md#initial-configuration](https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/README.md#initial-configuration) (after **_unclassified_** screenshot, before **_Examples_** section). Such a note could have a reference to the [official YAML Node Anchors](https://yaml.org/spec/1.2/spec.html#id2785586) documentation and a few links to the existing [JCasC demos](https://github.com/jenkinsci/configuration-as-code-plugin/tree/master/demos) (have to be adjusted as well). One could be easily taken from your test here or from [my GH issue](https://github.com/jenkinsci/configuration-as-code-plugin/issues/1246#issue-548497987). Probably, yet another reasonable use of such functionality with anchors might be for [permanent build agents](https://github.com/jenkinsci/configuration-as-code-plugin/issues/1246#issue-548497987), but practically speaking it's pretty much the same as other _cloud-templates_, either `eC2Fleet` or `kubernetes`...

**Comment by v1v** [comment]
neat! From the docs point of view, in the main README.md, I think it could be added as an example of using the node anchors, to be automatically validated with the annotation: `@ConfiguredWithReadme("README.md#id")` as done in the https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/integrations/src/test/java/io/jenkins/plugins/casc/TopReadmeTest.java.

**Comment by jetersen** [comment]
as a FYI, we use snakeyaml at the moment which is an parser for YAML 1.1 spec.

**Comment by jetersen** [comment]
I would like to avoid linking to the yaml spec as we do not follow the spec 100% for instance we do not support maps.

**Comment by jetersen** [comment]
As this has full test coverage and blocking #1375 I am going ahead and merging it.

**Pull request merged**